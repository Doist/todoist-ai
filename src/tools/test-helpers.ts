import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
