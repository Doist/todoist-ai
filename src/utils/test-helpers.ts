import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
