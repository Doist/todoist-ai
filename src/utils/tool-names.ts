/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
