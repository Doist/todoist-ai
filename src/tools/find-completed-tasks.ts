import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { mapTask } from '../tool-helpers.js'
import { ApiLimits } from '../utils/constants.js'
import { previewTasks, summarizeList } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_TASKS_BY_DATE, GET_OVERVIEW } = ToolNames

const ArgsSchema = {
    getBy: z
        .enum(['completion', 'due'])
        .default('completion')
        .describe(
            'The method to use to get the tasks: "completion" to get tasks by completion date (ie, when the task was actually completed), "due" to get tasks by due date (ie, when the task was due to be completed by).',
        ),
    since: z
        .string()
        .date()
        .regex(/^\d{4}-\d{2}-\d{2}$/)
        .describe('The start date to get the tasks for. Format: YYYY-MM-DD.'),
    until: z
        .string()
        .date()
        .regex(/^\d{4}-\d{2}-\d{2}$/)
        .describe('The start date to get the tasks for. Format: YYYY-MM-DD.'),
    workspaceId: z.string().optional().describe('The ID of the workspace to get the tasks for.'),
    projectId: z.string().optional().describe('The ID of the project to get the tasks for.'),
    sectionId: z.string().optional().describe('The ID of the section to get the tasks for.'),
    parentId: z.string().optional().describe('The ID of the parent task to get the tasks for.'),
    limit: z
        .number()
        .int()
        .min(1)
        .max(ApiLimits.COMPLETED_TASKS_MAX)
        .default(ApiLimits.COMPLETED_TASKS_DEFAULT)
        .describe('The maximum number of tasks to return.'),
    cursor: z
        .string()
        .optional()
        .describe(
            'The cursor to get the next page of tasks (cursor is obtained from the previous call to this tool, with the same parameters).',
        ),
}

const findCompletedTasks = {
    name: ToolNames.FIND_COMPLETED_TASKS,
    description: 'Get completed tasks.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { getBy, ...rest } = args
        const { items, nextCursor } =
            getBy === 'completion'
                ? await client.getCompletedTasksByCompletionDate(rest)
                : await client.getCompletedTasksByDueDate(rest)
        const mappedTasks = items.map(mapTask)

        const textContent = generateTextContent({
            tasks: mappedTasks,
            args,
            nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks: mappedTasks,
                nextCursor,
                totalCount: mappedTasks.length,
                hasMore: Boolean(nextCursor),
                appliedFilters: args,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    tasks,
    args,
    nextCursor,
}: {
    tasks: ReturnType<typeof mapTask>[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
}) {
    // Generate subject description
    const getByText = args.getBy === 'completion' ? 'completed' : 'due'
    const subject = `Completed tasks (by ${getByText} date)`

    // Generate filter hints
    const filterHints: string[] = []
    filterHints.push(`${getByText} date: ${args.since} to ${args.until}`)
    if (args.projectId) filterHints.push(`project: ${args.projectId}`)
    if (args.sectionId) filterHints.push(`section: ${args.sectionId}`)
    if (args.parentId) filterHints.push(`parent: ${args.parentId}`)
    if (args.workspaceId) filterHints.push(`workspace: ${args.workspaceId}`)

    // Generate helpful suggestions for empty results
    const zeroReasonHints: string[] = []
    if (tasks.length === 0) {
        zeroReasonHints.push('No tasks completed in this date range')
        zeroReasonHints.push('Try expanding the date range')
        if (args.projectId || args.sectionId || args.parentId) {
            zeroReasonHints.push('Try removing project/section/parent filters')
        }
        if (args.getBy === 'due') {
            zeroReasonHints.push('Try switching to "completion" date instead')
        }
    }

    // Generate contextual next steps
    const nextSteps: string[] = []
    if (tasks.length > 0) {
        nextSteps.push(
            `Use ${FIND_TASKS_BY_DATE} for active tasks or ${GET_OVERVIEW} for current productivity.`,
        )
        if (tasks.some((task) => task.recurring)) {
            nextSteps.push('Recurring tasks will automatically create new instances.')
        }
    }

    return summarizeList({
        subject,
        count: tasks.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewTasks(tasks),
        zeroReasonHints,
        nextSteps,
    })
}

export { findCompletedTasks }
