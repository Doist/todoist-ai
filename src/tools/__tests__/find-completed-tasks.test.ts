import type { Task, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { createMockTask, extractTextContent } from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findCompletedTasks } from '../find-completed-tasks.js'

// Mock the Todoist API
const mockTodoistApi = {
    getCompletedTasksByCompletionDate: jest.fn(),
    getCompletedTasksByDueDate: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_COMPLETED_TASKS } = ToolNames

describe(`${FIND_COMPLETED_TASKS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('getting completed tasks by completion date (default)', () => {
        it('should get completed tasks by completion date', async () => {
            const mockCompletedTasks: Task[] = [
                createMockTask({
                    id: '8485093748',
                    content: 'Completed task 1',
                    description: 'Task completed yesterday',
                    completedAt: '2024-01-01T00:00:00Z',
                    labels: ['work'],
                    priority: 2,
                    url: 'https://todoist.com/showTask?id=8485093748',
                    addedAt: '2025-08-13T22:09:56.123456Z',
                    due: {
                        date: '2025-08-14',
                        isRecurring: false,
                        lang: 'en',
                        string: 'Aug 14',
                        timezone: null,
                    },
                }),
            ]

            mockTodoistApi.getCompletedTasksByCompletionDate.mockResolvedValue({
                items: mockCompletedTasks,
                nextCursor: null,
            })

            const result = await findCompletedTasks.execute(
                {
                    getBy: 'completion',
                    limit: 50,
                    since: '2025-08-10',
                    until: '2025-08-15',
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getCompletedTasksByCompletionDate).toHaveBeenCalledWith({
                since: '2025-08-10',
                until: '2025-08-15',
                limit: 50,
            })

            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle explicit completion date query', async () => {
            mockTodoistApi.getCompletedTasksByCompletionDate.mockResolvedValue({
                items: [],
                nextCursor: 'next-cursor',
            })

            const result = await findCompletedTasks.execute(
                {
                    getBy: 'completion',
                    limit: 100,
                    since: '2025-08-01',
                    until: '2025-08-31',
                    projectId: 'specific-project-id',
                    cursor: 'current-cursor',
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getCompletedTasksByCompletionDate).toHaveBeenCalledWith({
                since: '2025-08-01',
                until: '2025-08-31',
                projectId: 'specific-project-id',
                limit: 100,
                cursor: 'current-cursor',
            })

            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('getting completed tasks by due date', () => {
        it('should get completed tasks by due date', async () => {
            const mockCompletedTasks: Task[] = [
                createMockTask({
                    id: '8485093750',
                    content: 'Task completed by due date',
                    description: 'This task was due and completed',
                    completedAt: '2024-01-01T00:00:00Z',
                    labels: ['urgent'],
                    priority: 3,
                    url: 'https://todoist.com/showTask?id=8485093750',
                    addedAt: '2025-08-13T22:09:58.123456Z',
                    due: {
                        date: '2025-08-15',
                        isRecurring: true,
                        lang: 'en',
                        string: 'every Monday',
                        timezone: null,
                    },
                }),
            ]

            mockTodoistApi.getCompletedTasksByDueDate.mockResolvedValue({
                items: mockCompletedTasks,
                nextCursor: null,
            })

            const result = await findCompletedTasks.execute(
                {
                    getBy: 'due',
                    limit: 50,
                    since: '2025-08-10',
                    until: '2025-08-20',
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getCompletedTasksByDueDate).toHaveBeenCalledWith({
                since: '2025-08-10',
                until: '2025-08-20',
                limit: 50,
            })
            expect(mockTodoistApi.getCompletedTasksByCompletionDate).not.toHaveBeenCalled()

            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('label filtering', () => {
        it.each([
            {
                name: 'single label with OR operator',
                params: {
                    getBy: 'completion' as const,
                    since: '2025-08-01',
                    until: '2025-08-31',
                    limit: 50,
                    labels: ['@work'],
                    labelsOperator: 'or' as const,
                },
                expectedMethod: 'getCompletedTasksByCompletionDate',
                expectedFilter: '(@work)',
            },
            {
                name: 'multiple labels with AND operator',
                params: {
                    getBy: 'due' as const,
                    since: '2025-08-01',
                    until: '2025-08-31',
                    limit: 50,
                    labels: ['@work', '@urgent'],
                    labelsOperator: 'and' as const,
                },
                expectedMethod: 'getCompletedTasksByDueDate',
                expectedFilter: '(@work  &  @urgent)',
            },
            {
                name: 'multiple labels with OR operator',
                params: {
                    getBy: 'completion' as const,
                    since: '2025-08-10',
                    until: '2025-08-20',
                    limit: 25,
                    labels: ['@personal', '@shopping'],
                    labelsOperator: 'or' as const,
                },
                expectedMethod: 'getCompletedTasksByCompletionDate',
                expectedFilter: '(@personal  |  @shopping)',
            },
        ])(
            'should filter completed tasks by labels: $name',
            async ({ params, expectedMethod, expectedFilter }) => {
                const mockCompletedTasks = [
                    createMockTask({
                        id: '8485093748',
                        content: 'Completed task with label',
                        labels: params.labels,
                        completedAt: '2024-01-01T00:00:00Z',
                    }),
                ]

                const mockResponse = { items: mockCompletedTasks, nextCursor: null }
                const mockMethod = mockTodoistApi[
                    expectedMethod as keyof typeof mockTodoistApi
                ] as jest.MockedFunction<
                    (...args: never[]) => Promise<{ items: unknown[]; nextCursor: string | null }>
                >
                mockMethod.mockResolvedValue(mockResponse)

                const result = await findCompletedTasks.execute(params, mockTodoistApi)

                expect(mockMethod).toHaveBeenCalledWith({
                    since: params.since,
                    until: params.until,
                    limit: params.limit,
                    filterQuery: expectedFilter,
                    filterLang: 'en',
                })

                const textContent = extractTextContent(result)
                expect(textContent).toMatchSnapshot()
            },
        )

        it('should handle empty labels array', async () => {
            const params = {
                getBy: 'completion' as const,
                since: '2025-08-01',
                until: '2025-08-31',
                limit: 50,
                labels: [],
                labelsOperator: 'or' as const,
            }

            const mockResponse = { items: [], nextCursor: null }
            mockTodoistApi.getCompletedTasksByCompletionDate.mockResolvedValue(mockResponse)

            await findCompletedTasks.execute(params, mockTodoistApi)

            expect(mockTodoistApi.getCompletedTasksByCompletionDate).toHaveBeenCalledWith({
                since: params.since,
                until: params.until,
                limit: params.limit,
            })
        })

        it('should combine other filters with label filters', async () => {
            const params = {
                getBy: 'due' as const,
                since: '2025-08-01',
                until: '2025-08-31',
                limit: 25,
                projectId: 'test-project-id',
                sectionId: 'test-section-id',
                labels: ['@important'],
                labelsOperator: 'or' as const,
            }

            const mockTasks = [
                createMockTask({
                    content: 'Important completed task',
                    labels: ['important'],
                    completedAt: '2024-01-01T00:00:00Z',
                }),
            ]
            const mockResponse = { items: mockTasks, nextCursor: null }
            mockTodoistApi.getCompletedTasksByDueDate.mockResolvedValue(mockResponse)

            const result = await findCompletedTasks.execute(params, mockTodoistApi)

            expect(mockTodoistApi.getCompletedTasksByDueDate).toHaveBeenCalledWith({
                since: params.since,
                until: params.until,
                limit: params.limit,
                projectId: params.projectId,
                sectionId: params.sectionId,
                filterQuery: '(@important)',
                filterLang: 'en',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
        })
    })

    describe('error handling', () => {
        it('should propagate completion date API errors', async () => {
            const apiError = new Error('API Error: Invalid date range')
            mockTodoistApi.getCompletedTasksByCompletionDate.mockRejectedValue(apiError)

            await expect(
                findCompletedTasks.execute(
                    // invalid date range
                    {
                        getBy: 'completion',
                        limit: 50,
                        since: '2025-08-31',
                        until: '2025-08-01',
                        labels: [],
                        labelsOperator: 'or' as const,
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Invalid date range')
        })

        it('should propagate due date API errors', async () => {
            const apiError = new Error('API Error: Project not found')
            mockTodoistApi.getCompletedTasksByDueDate.mockRejectedValue(apiError)

            await expect(
                findCompletedTasks.execute(
                    {
                        getBy: 'due',
                        limit: 50,
                        since: '2025-08-01',
                        until: '2025-08-31',
                        projectId: 'non-existent-project',
                        labels: [],
                        labelsOperator: 'or',
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Project not found')
        })
    })
})
