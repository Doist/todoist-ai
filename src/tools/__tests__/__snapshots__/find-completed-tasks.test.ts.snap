// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-completed-tasks tool getting completed tasks by completion date (default) should get completed tasks by completion date 1`] = `
"Completed tasks (by completed date): 1 (limit 50).
Filter: completed date: 2025-08-10 to 2025-08-15.
Preview:
    Completed task 1 • due 2025-08-14 • P3 • id=8485093748
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity."
`;

exports[`find-completed-tasks tool getting completed tasks by completion date (default) should handle explicit completion date query 1`] = `
"Completed tasks (by completed date): 0 (limit 100), more available.
Filter: completed date: 2025-08-01 to 2025-08-31; project: specific-project-id.
No results. No tasks completed in this date range; Try expanding the date range; Try removing project/section/parent filters.
Next:
- Pass cursor 'next-cursor' to fetch more results."
`;

exports[`find-completed-tasks tool getting completed tasks by due date should get completed tasks by due date 1`] = `
"Completed tasks (by due date): 1 (limit 50).
Filter: due date: 2025-08-10 to 2025-08-20.
Preview:
    Task completed by due date • due 2025-08-15 • P2 • id=8485093750
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity.
- Recurring tasks will automatically create new instances."
`;

exports[`find-completed-tasks tool label filtering should combine other filters with label filters 1`] = `
"Completed tasks (by due date): 1 (limit 25).
Filter: due date: 2025-08-01 to 2025-08-31; project: test-project-id; section: test-section-id; labels: @important.
Preview:
    Important completed task • P4 • id=8485093748
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity."
`;

exports[`find-completed-tasks tool label filtering should filter completed tasks by labels: multiple labels with AND operator 1`] = `
"Completed tasks (by due date): 1 (limit 50).
Filter: due date: 2025-08-01 to 2025-08-31; labels: @work & @urgent.
Preview:
    Completed task with label • P4 • id=8485093748
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity."
`;

exports[`find-completed-tasks tool label filtering should filter completed tasks by labels: multiple labels with OR operator 1`] = `
"Completed tasks (by completed date): 1 (limit 25).
Filter: completed date: 2025-08-10 to 2025-08-20; labels: @personal | @shopping.
Preview:
    Completed task with label • P4 • id=8485093748
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity."
`;

exports[`find-completed-tasks tool label filtering should filter completed tasks by labels: single label with OR operator 1`] = `
"Completed tasks (by completed date): 1 (limit 50).
Filter: completed date: 2025-08-01 to 2025-08-31; labels: @work.
Preview:
    Completed task with label • P4 • id=8485093748
Next:
- Use find-tasks-by-date for active tasks or get-overview for current productivity."
`;
